### ISSUE TITLE
[discussion] integrating nan or similar with io.js

### ISSUE THREAD

**Issue by trevnorris** [iojs/nan#349]
nan has done wonders for the community, but there are still missing bits. If these can be addressed there is a good desire to bring this into io.js itself. This thread should serve as a discussion on what are all the missing bits, what we can do about them and how to eventually get this into io.js.

The most prominent issue is lack of ABI compatibility. Needing to recompile with every new release, especially now with more frequent V8 updates, is difficult. @bnoordhuis mentioned that using templates is not a good form of ABI compatibility because they don't play well with the linker. (Ben, if you have any additional thoughts on this please feel free to include anything I've missed).

**Comment by kkoopa** [comment]
I'll start by noting that not using templates leads to very clumsy code, and since V8 makes heavy use of templates, this would require not directly exposing anything from V8, otherwise the issue of ABI compatibility will reemerge.

**Comment by rvagg** [comment]
Some initial randomish thoughts:

1. This is a separate effort to NAN and will be until it can completely replace NAN and we can deprecate it - even then we'll probably have overlap while people adjust
2. Let's start simple and unambitious
3. We're going to have to accept a world similar to NAN where the API/ABI will have to change over time to adapt to a changing upstream API but the benefit will be that you're dealing with a single API to code against
4. People are going to want to reach around to V8 regardless of how sophisticated this is, we need to accept that and possibly make allowances for it, the problem with https://github.com/tjfontaine/node-addon-layer as I see it is that it goes too far in one leap and makes no allowances for anything but standard uses (the typical "framework problem"); this and item (2) above are the main reasons I've advocated for the NAN approach over the addon-layer approach in the past
5. The continued instability from V8 is getting me down, they seem to have an endless appetite to screw with their API and therefore screw users of the C++ API, maybe it's the frustration with having to deal with permanent backward compatibility in JS to ES1 that forces them to take it out on the C++ API. @domenic says he's going to take this up with the V8 team but at this stage I'm much more willing to support efforts to detach Node from V8 as a direct dependency.
6. Chakra introduces an interesting dynamic here and we may want to engage the Microsoft team on this effort given what they've had to go through to make their shim

**Comment by trevnorris** [comment]
Three things I believe could be done without templates and be ABI compliant:

1. `Local<Value>` with a few methods. Such as `Is{Undefined,Null,True,etc.}()`, `BooleanValue()`, `NumberValue()`, etc. The hurdle to overcome would be casting to other types, and supporting all those APIs.
2. `typedef void (*FunctionCallback)(const FunctionCallbackInfo<Value>& info)` could be a normal function call. `info[i]` could be an array of values, and since they're all initially `Local<Values>` the naive implementation would be supported. The problematic would be all calls that return `Local<Object>`.
3. Buffers are pretty straight forward. This should even be possible using our `Local<Value>` abstraction, and in fact would be most compatible that way because currently it requires a `Local<Object>` and in the future will require a `Local<Uint8Array>`. Possibly something like the following (warning, this is very C-ish):
```cpp
struct node_buffer_s {
  char* data;
  size_t length;
};
typedef void node_buffer_s node_buffer_t;
void FnCallback(const CallbackInfo* info) {
  node_buffer_t buf;
  Node::Buffer::GetData(info[0], &buf);
}
```

**Comment by kkoopa** [comment]
Would definitely want a code generator to generate the code for all the ensuing virtually identical wrapper functions with different types. It would be nice if one could get at the resulting C++ code after template processing, but before compilation.

Have a look at how Python 3 is doing it. Python is also dynamically typed, and everything is an object, but I have to say that writing addons for it is rather painful.

Also, I think the bar for a useful addon API should be set higher than trivial. If you can only do trivial things, you might as well just set up a socket or pipe of sorts and communicate between JavaScript and the real world throught that.

**Comment by trevnorris** [comment]
@kkoopa 
> Also, I think the bar for a useful addon API should be set higher than trivial.

I was just echoing @rvagg's sentiment of:

> Let's start simple and unambitious

So to clarify I was simply saying we start small, experiment and see how well it works. If it works at all.

One important thing is that if we want to truly achieve ABI compatibility then the exposed API needs to be written in C. There have been more than a few people smarter than myself that have expressed that C++ doesn't do well at maintaining proper memory structures when compiled with different compilers, or even different versions of the same compiler.

**Comment by kkoopa** [comment]
I agree that we have to start simple, implementation-wise. However, that initial implementation needs to scale up to offering a richer feature set.

It does not need to be written in C as such, it's just the external functions (and data) that need to be "C" exported. Without C++, how would the the wrappers interact with v8?

**Comment by trevnorris** [comment]
> It does not need to be written in C as such, it's just the external functions (and data) that need to be "C" exported.

Agreed. I should have made my following statement more clear:

> One important thing is that if we want to truly achieve ABI compatibility then **the exposed API** needs to be written in C.

So yes. I completely agree. It's just the user facing API that should be C.

**Comment by kkoopa** [comment]
Something like this should do.
https://github.com/martine/v8c/blob/v8c/src/v8c.cc

**Comment by trevnorris** [comment]
@kkoopa good stuff. we'd just have to extend it to support current APIs like `Buffer`.

**Comment by kkoopa** [comment]
And new V8. Notice that this project has been abandoned for 6 years.

**Comment by bnoordhuis** [comment]
> One important thing is that if we want to truly achieve ABI compatibility then the exposed API needs to be written in C. There have been more than a few people smarter than myself that have expressed that C++ doesn't do well at maintaining proper memory structures when compiled with different compilers, or even different versions of the same compiler.

Structures/classes don't have to be a problem when you don't use features such as virtual methods, multiple inheritance, member function pointers, etc.

I suppose name mangling could be an issue for people that mix C++11 code with pre-C++11 code.  You can work around it by building everything with `-fabi-version=6` but that requires that you use a recent compiler.

**Comment by kkoopa** [comment]
So, what we'd essentially want is to wrap v8's C++ exports in C exports to get a stable ABI and then have a (NAN-derived) header library on top of that which would restore the C++ nature of V8's API by mapping to the C exports.

Something like this:
```c++
extern "C" {
  V8StringUtf8 v8_string_utf8_new(const char* data, int length);
  V8Number v8_number_int32_new(int32_t val);
  V8Number v8_number_uint32_new(uint32_t val);
}
```
```c++
template<v8::String>
v8::Local<v8::String> NanNew(const char *data, int length = -1) {
  return v8_string_utf8_new(data, length);
}

template<v8::Int32>
v8::Local<v8::Int32> NanNew(int32_t val) {
  return v8_number_int32_new(val);
}

template<v8::UInt32>
v8::Local<v8::UInt32> NanNew(uint32_t val) {
  return v8_number_uint32_new(val);
}
```

**Comment by kzc** [comment]
Microsoft appears to have created a V8 compatible C++ API facade for Chakra. While it is a nice piece of engineering, it also has the disadvantage of trying to keep up with a constantly changing V8 API that is not managed by the node project.

I like the javascript engine wrapper C API idea proposed by @trevnorris.  Node itself and node native modules could only use this C API. That way the javascript engine (V8, Spidermonkey, Duktape, JavascriptCore or other) could be made to be pluggable at runtime depending on the needs of the user. Native node modules compiled against this hypothetical C API could work against any such runtime-pluggable engine without recompilation due to its more stable ABI.

However, one downside with such a javascript engine C API would be that it would impose a little bit of overhead on every javascript <=> c++ operation as compared to header-only inline C++ classes or C macros. Even so, I still think it is worth the cost for the maintainability and the flexibility it would offer.

**Comment by trevnorris** [comment]
@kzc After doing testing in this area I'm fairly confident that we'd be able to do this with very little overhead. At least minimal enough where users wouldn't feel it in their JS code.

**Comment by kkoopa** [comment]
Clearly yes. The overhead of context switching from JavaScript to C++ drowns 
out everything else.

**Comment by kzc** [comment]
I wouldn't recommend having the proposed C API functions and types mirror v8's C++ API and its concepts, because you would still have the original problem of tightly coupling node against an ever changing v8. Ideally the proposed C API should work equally well with Spidermonkey which cares about registering GC roots and Duktape which uses a Lua-style sandbox approach where you only manipulate copies of the actual data structures. But I recognize that such portability would impose a runtime cost.

**Comment by kkoopa** [comment]
On the contrary, I would recommend having it mirror V8's C++ API. Abstracting away the engine in a pluggable way is too much unnecessary work. It is not like io.js core would use these bindings internally, so it still would not become Spidernode. Personally, I have zero interest in supporting other JavaScript engines than V8.

**Comment by kzc** [comment]
Perhaps I mistook @trevnorris' proposal. I thought both node itself and native node modules would be changed to only use this new C API to isolate them both from the javascript engine's API.

If the goal is to only support v8, then I don't think the proposed C API buys you much. By closely mirroring the C++ v8 API any significant upstream change would still require an equivalent change in the C API.

**Comment by trevnorris** [comment]
@kzc The gain for maintaining a public C API is that native modules will not have to be recompiled every time they upgrade their version of node.

@kkoopa Unfortunately using this API internally is exactly what large businesses want. Microsoft has already done a port to Chakra, Oracle and IBM are independently working to get node running on Java. Anyway, just giving you a heads up that this is exactly what these large companies are after.

**Comment by kkoopa** [comment]
Oh, really? Why would they want that? What, exactly, are we discussing here? Is it making an ABI-stable API for modules or is it rewriting all of Node to abstract away V8? The former is a lot less work than the latter.

**Comment by trevnorris** [comment]
They want the ability to use whatever VM under the hood they deem appropriate. But then also have the user facing API ABI stable so they can ship pre-compiled modules. I'm only relaying this information so you're aware, and I dread to think what would happen to core in an attempt to do this.

The reason I opened this was to discuss the **user-facing** part. Creating ABI stability for native modules so they don't have to be recompiled with every release.

**Comment by kkoopa** [comment]
Good, then we're on the same page. In that case, I think the best to do is the following:

Take (almost) all the functionality already exposed by NAN (diregarding sugar, 96 % of it is used in native modules in the wild). This is the least a usable API should offer to support all the existing native modules.

Create C-exported wrappers for all the necessary V8 functions. Naming should reflect the original API, so let's use some arbitrary, but consistent, sort of mangling system through underscores.
The v8 namespace may be considered implicit. We only expose the longest function signature when there are convenience overloads or default argument values.
```c++
Local__String__ String_NewFromUtf8(Isolate *, const char *str, int length, ...) {...}
Local__String__ String_NewFromTwoByte(Isolate *, const uint16_t *str, int length, ...) {...}
...
```
Then, as I mentioned in an earlier post, we could stick a header-only C++ API back on top, e.g. NAN or derivative. This library would circumvent the clumsiness and loss from having C-bindings, giving the best of two worlds. A C++ API with consistent C ABI. The naming scheme for C-exported functions can be arbitrarily clumsy, yet still hidden away.

All of this even seems like it would be possible to automate, i.e. Write a program that parses v8.h, getting all public functions and spits out a bunch of horribly named C function declarations and a C++ header that reconstructs the C++ API via the C functions. This would be cool, but probably not an easy task.

**Comment by trevnorris** [comment]
Sounds like a good enough approach to me. @bnoordhuis thoughts?

**Comment by kkoopa** [comment]
Would still need to export some functions from Node too, buffer, objectwrap, strings, etc. Then have som representation of exported classes as structs (they are rare,but a couple exist).

**Comment by bnoordhuis** [comment]
Something to consider: V8 uses RAII in a few places to good effect, c.f. HandleScope.  A C shim would be quite error prone to use.

Exposing structs in the API might make ABI stability pretty complicated.  (If I ever get around to writing a blog post "Libuv, lessons learned", that's going to feature prominently.)  I would suggest using opaque pointers, coupled with accessors.

**Comment by kkoopa** [comment]
Can you give some concrete examples of what you had in mind, e.g. HandleScope?

**Comment by kzc** [comment]
@trevnorris, I agree that a C API wrapper makes sense if you intend to support javascript engines other than just v8.  Just be aware that there are features and concepts in the v8 API that don't easily map to other javascript engines, and vice versa. Mirroring the v8 C++ API in C may create more work for the other engine mappings.

**Comment by trevnorris** [comment]
@kzc At this point I'm neither for or against mirroring the V8 API. Right now I'm just trying to facilitate conversation until we can achieve some actionable items.

**Comment by bnoordhuis** [comment]
> Can you give some concrete examples of what you had in mind, e.g. HandleScope?

I think that's a question for me?  Well, I imagine that C++ code like this:

```c++
void f(v8::Isolate* isolate) {
  v8::HandleScope handle_scope(isolate);
  // ...
  if (g()) return;
  // ...
}
```

Would end up looking something like this in C:

```c
void f(v8_isolate *isolate) {
  v8_handlescope_enter(isolate);
  // ...
  if (g()) {
    v8_handlescope_exit(isolate);
    return;
  }
  // ...
  v8_handlescope_exit(isolate);
}
```

Manually having to balance enter/exit calls everywhere is a bit of a pain and easy to get wrong.

**Comment by trevnorris** [comment]
@bnoordhuis so the best solution would be to still use a C++ interface but stay away from (as you mentioned) virtual methods, multiple inheritance, member function pointers, etc?

**Comment by bnoordhuis** [comment]
I don't know about 'best' but a C++ API done right would certainly be easier and more robust to use.

**Comment by kkoopa** [comment]
Note that the idea was to reconstruct the C++ API on top of the C API. But, if it is less work and doable with a C++ API, then I guess that would be better.

**Comment by orangemocha** [comment]
+1 on achieving ABI compatibility. 

I think that it would be hard to achieve ABI compatibility in C++. Here are two issues that I am aware of:
1) If the class definition changes even just by adding a private field at the end, the size of objects changes and any use of the new operator is broken across module boundaries. I think you could work around this by allowing instantiation only in the main module (node) through factory methods, and disallowing any passing by value in the addon module (because the copy constructor would be affected too).
2) Name mangling. This can break across compiler versions. I am not aware of any silver bullets to address this issue.

**Comment by orangemocha** [comment]
> Chakra introduces an interesting dynamic here and we may want to engage the Microsoft team on this effort given what they've had to go through to make their shim

cc @geoffkizer, @jianchun, @curtisman

**Comment by bnoordhuis** [comment]
> 1) If the class definition changes even just by adding a private field at the end, the size of objects changes and any use of the new operator is broken across module boundaries. I think you could work around this by allowing instantiation only in the main module (node) through factory methods, and disallowing any passing by value in the addon module (because the copy constructor would be affected too).

That's correct.  The same issue applies to C structs, however.  I don't think C++ is any worse in that respect, it's just different.

> 2) Name mangling. This can break across compiler versions. I am not aware of any silver bullets to address this issue.

With gcc and clang, you control that through the `-fabi-version=` switch.  I don't know if VS has a similar option.

The C++ ABI changes very infrequently.  Before C++11, the last real update was in 2004, I think.

**Comment by orangemocha** [comment]
>The same issue applies to C structs, however. I don't think C++ is any worse in that respect, it's just different.

True. I suppose you could design a C++ API that provides ABI compatibility. As you mentioned, it would have to disallow a few features that are specific to C++.

Another advantage of using C is, well, that you can invoke it from pure C code. It’s the lowest common denominator, on top of which you can build bindings for other languages.

So I am +1 for the approach (suggested by @kkoopa) of building the syntactic C++ sugar on top of a flat C layer, which would help streamline things like balancing the enter/exit scope calls.

>With gcc and clang, you control that through the -fabi-version= switch. I don't know if VS has a similar option.

I’ll try to find out more about name mangling stability in VC++.

Aside from the C vs C++ issue, I think that in order to achieve ABI compatibility the API needs to be designed and maintained with that goal in mind. As @kzc pointed out in [this comment](https://github.com/nodejs/nan/issues/349#issuecomment-102416977) that means that we shouldn’t expose v8 types directly, or we would be back to square one. v8 types can be encapsulated, or passed as opaque tokens (e.g. void *) but modules should not be exposed to their definitions. Correct?


Also, I think it makes a lot of sense to consider the interface between Node itself and the engine as part of this conversation. Please correct me if I am wrong, but an internal abstraction between Node and v8 would also solve the problem of native modules interfacing with Node, while the other way around is not necessarily true. Well, because the former is probably a bigger beast to tackle than the latter. It would also mitigate the pain of Node core keeping up with v8, and possibly allow for pluggable engines. My point is that at the time that we introduce a new API for native addons, and break compatibility, we might want to consider solving the other problems as well, because we don’t want to change this API again in 6 months. API should be designed to last. I guess that’s an argument against starting small and unambitious ;)

On that topic, it should be noted that JxCore has already implemented an abstraction layer that encapsulates the engine interface for both “Node” core and native modules. They have done so while maintaining compatibility with existing native modules written against v8. And they have made all the change in core to interface with this abstracted engine API. I have had the opportunity to talk with the folks who worked on the project, and they seemed willing and able to extricate the interface changes from the other features and submit a pull request to Node. Caveats: their changes are currently based on Node v0.10; and I don’t think their API is 100% ABI-ready in that it uses C++ and macros, but fixing that might be a fairly mechanical job.

So I think one possible course of action to address this issue at large could be the following:
1.  Engage JxCore’s help to….
2.  Port their changes from 0.10 to “converged node” master. This would also be a good test to see how much their abstraction layer is able to abstract v8 from old versions to the recent versions.
3.  Revisit their API to be more ABI-friendly, substituting macros with actual functions. Consider a mapping to C.
4.  Open a PR to Node

Does that sound like a viable option?

**Comment by bnoordhuis** [comment]
I'm personally highly skeptical that any kind of engine abstraction is workable in the long run and not a huge time drain and source of friction.

JXCore basically shims the V8 API, similar to the Chakra port.  Updating V8 can already be big source of friction; now factor in the cost of updating all the shims.

What if you want to use V8-specific functionality?  You can't unless the other engines have similar functionality.

What if the V8 API changes in a way that you can't reasonably emulate with other engines?  Or vice versa?  Do we stop upgrading?

Let JXCore and Chakra show that their approach is viable over a multi-year time span first before we start thinking about adopting it.

**Comment by pittek** [comment]
I've been recently walking through JXCore sources. Important points to remark;

 1. V8 and SpiderMonkey Javascript engines are already 180 degree different. (Native type incompatibilities, GCRooting, Concurrent GC, No GC in GC, Runtime is not an isolate,  lot more)
 2. Emulation is not an option. JXCore doesn't emulate V8. My impression was only the final macro interface pretends like that. 
 3. Same interface also covers multiple v8 isolates yet I'm not sure why they use multiple isolates ( I didn't walk through the docs yet) 

@trevnorris clearly C,C++ etc. wrapper is not a solution alone. I agree with @orangemocha on JXCore's engagement in order to make the process less painful. 

`What if the V8 API changes in a way that you can't reasonably emulate with other engines? Or vice versa? Do we stop upgrading?`

@bnoordhuis I don't know if you are familiar with SpiderMonkey. V8 and SpiderMonkey are already in a place that you can not reasonably emulate. 

`What if you want to use V8-specific functionality?`
@bnoordhuis can you be more specific? Sorry if I'm wrong but Does Node.JS promotes OS specific features ? Besides, other JavaScript engines also have their unique advantages.  

One suggestion though. nan and jxcore can collaborate on a common solution for c,c++ node modules. This could be a good starting point and community would benefit from that.

**Comment by trevnorris** [comment]
@pittek OS specific != V8 specific. The specifics come when doing things like tooling around the VM, or even things like how we attach C++ class instances to Objects, and the Object to the class instances.

**Comment by pittek** [comment]
@trevnorris actually OS specific = V8 specific if you are exactly referring to tooling around the VM. How VM is different than OS ? Attaching C++ class instances to `Objects?` is the least concern that an abstraction would take care of. However how you manage the life cycle of that Object / Scope and Runtime is another thing. For example, LibUV plays some kind of an abstraction role between the underlying OS and Node.JS.

**Comment by kkoopa** [comment]
V8 is a library, not an operating system. They are quite different. V8 runs on many operating systems, so does Node, which depends on V8. Here, libuv abstracts away the operating system, with varying degrees of success.

**Comment by plika** [comment]
IMHO, there is no single best pattern that works for each JavaScript engine interface and internals. JXcore encapsulates SpiderMonkey while abstracting V8. Chakra is much more suitable for abstraction but in time the differences between the engines may require encapsulation taking the place.  

I believe the real problem is something else. 

The problem here is abstracting JavaScript engine capabilities to 3rd party solutions. JXcore has a separate abstraction for the JavaScript engine features. It doesn't have much to do how we abstract or encapsulate the JavaScript engine. It's all about sharing a common interface. 

Today we use this interface for iOS and Android applications mostly. For example on Windows, once you have JXcore SM or V8 build as a shared library, feel free to switch it without recompiling the host application. 

This interface is usable for the majority of c,c++ module scenarios. We can still allow old  style module development but don't encourage it. I don't think a module developer would pick an option that will end up updating again and again.  

As far as I know, majority of the c,c++ modules doesn't require a special access to core JavaScript engine. 

Any thoughts ?

**Comment by obastemur** [comment]
Hopefully it is not going to be a wrong statement but I consider io.js as the node.js next. So, what will be the next is very important in terms of bringing a rather consistent interface for addons. (c, c++ addons)

I see node.js is engine independent framework in the future. However, node serving on a multiple engine interface is not that easy as a short/mid term goal. I agree with @bnoordhuis at some level that the multiple engine approach should evolve. But this process is not going to evolve alone.

IMHO, it's really nice to have all these (jxcore, chakra) efforts around node.js. I don't see a reason why we shouldn't combine the efforts. We can start with a native addon interface that will be backing the next gen node applications. My humble addition to @rvagg's initial list;

 - The full API consists of a Compact  + Platform + Extended APIs 
 - A developer may assume that the compact API will be working on every platform and JS backend
 - Platform API shims the JS engine specific features with a promise of recompiling should work at some level.
 - Extended API provides the tools that we can interact directly with the JS backend. No promises.
 - Provide a template that a node addon can be integrated on run time and/or compile time. (shared / static library)

I believe, If we can establish a proper native addon interface, it would open a door to multiple JS engine backend support.

**Comment by kkoopa** [comment]
> ...I consider io.js as the node.js next

Yes, [that is sort of how it is](https://github.com/nodejs/dev-policy#repository-reconciliation-plan).
```
                        now
       (io.js)    v1.6   :  v1.7    v1.x
          |         |    :    |       |
 v0.10.x  /--------------:-----------------\   Node.js 2.0
____|____/               :                  \______|_____
         \               :                  /
          \--------------:-----------------/
          |         |    :     |       |
       (node.js) v0.12.x :  v0.13.x  v0.14.x
```
> The full API consists of a Compact + Platform + Extended APIs

That is sort of [how it is intended](https://github.com/nodejs/dev-policy#stability-policy) already.
```
   +---------------------------------------------------+
   |          Module and Application Ecosystem         |
   +---------------------------------------------------+
        |                |            |              |
        |                | +----------------------+  |
        |                | |  Binary Abstraction  |  |
        |                | |        Layer         |  |
        |                | +----------------------+  |
        |                |     |             |       |
   +----------------+    |     |             |       /
   |  Node.js Core  |    |     |             |      /
   |  Library API   |    |     |             |     /
   +----------------+    |     |             |    /
   |     js impl    |    |     |             |   /
   +----------------+    |     |             |  /
            |            |     |             | /
   +--------------------------------------+  |/
   | Node.js Application Binary Interface |  |
   +--------------------------------------|  |
   |              C/C++ impl              |  |
   +--------------------------------------+  |
                      |                      |
        +---------------------------------------+
        | Dependencies: v8, libuv, openssl, etc |
        +---------------------------------------+
```

**Comment by obastemur** [comment]
@kkoopa thanks for sharing the schema. I think I do remember seeing this definition once.  Let me remark the differences over the schema then;

 - There we need a compact set of API from `Node.JS Core Library API` + `Binary Abstraction Layer`. Kindly consider this as the basics of being a cross platform framework. The 'Platform and Extended' parts of the API I was mentioning applies to the rest.
 - I would rather separate Module and Application Ecosystems initially. They both have common and different set of challenges. One minor example; calling JSON.stringify on an object that internally marked as callable is not ECMA compliant. Yet v8 lets you to do that and some node modules depend on it. Unfortunately today, this is not something node can easily control. In short, if we accomplish a success with Module Ecosystem, the Application Ecosystem would follow that at some level of success.
 - Hard to imagine this API without a light type system. This is kind of harder than it sounds.
 - Although the dependencies have separate set of challenges,  they do have the common problems in terms of compact API. For example: Let's say V8 forces a zLib dependency which is not compatible to Compact BAL...
 - Better keep direct access to dependencies optional. This would bring / ease alternative ways to compile / embed compact API based solutions on limited platforms. 

Sorry for the naive question; Is there a working group on this subject or how we/someone propose a working group ?

**Comment by kkoopa** [comment]
This seems to be the 'Addon API' working group. That means the *Binary Abstraction Layer*, and apparently the *Node.js Application Binary Interface*, because no existing WG would fit it better and they are sort of related.

Abstracting away V8 is sort of out of scope here. While I think everyone agrees that "it would be nice to have a wonderful abstraction layer for the JS engine", that is like having a desire for world peace: nice, but unfeasible. That being said, as there appears to be interest in spidernode or whatever, it seems reasonable to form a WG around that and let those who care for it work on it.

I don't know how new working groups are added. Do they have to go through a TC meeting, or is it just a matter of someone with admin rights creating a repository? 

FFI [will be brought to io.js](https://github.com/nodejs/io.js/pull/1750), which should make trivial addons doable straight from JavaScript. That fits in the *Core Library and API*.

**Comment by kzc** [comment]
I see JXCore's javascript engine abstraction API makes extensive use of macros to achieve native module source code portability between v8 and Spidermonkey with recompilation. Although Chakra doesn't appear to use macros in its Node facade, it also seems to be a compile-time binding. @obastemur, could you comment on the feasibilty and estimated amount of effort to produce a pure C wrapper approach in order to achieve a stable ABI for node modules independent of the JS engine?

**Comment by obastemur** [comment]
> could you comment on the feasibilty and estimated amount of effort to produce a pure C wrapper approach in order to achieve a stable ABI for node modules independent of the JS engine?

@kzc JXcore's public API for embedder application is a pure C wrapper (otherwise we wouldn't able to embed it into Objective-C app directly)  And it's engine independent. Although our intention wasn't providing something for addons, we can update it to cover this part. In other words, this task has already been done for a compact set of JS engine features. It's not a rocket science what we did and the limited set of API can be extended. 

referencing from @plika's comment;
> For example on Windows, once you have JXcore SM or V8 build as a shared library, feel free to switch it without recompiling the host application. 

@kkoopa 
> Abstracting away V8 is sort of out of scope here. 

Let's say v8 version X is either a starting or end point. Considering the differences let say SpiderMonkey is Y. Preparing a `compact API` from X to Y (vice versa) doesn't abstract the V8 away. Instead, it helps to design something less vulnerable to X to Y or Y to X changes. At the JS engine basics level (compact API) it is doable and covers majority of the use cases.

> I think everyone agrees that "it would be nice to have a wonderful abstraction layer for the JS engine", that is like having a desire for world peace

It's rather a sampling problem. If we can pick the right sample from the given group we may produce something that makes sense. I wouldn't also go for something like 'for once for all'.. For me, this is not a 'let's replace v8 to something' problem. We already did that. I'm ready to help if you want to do the same and also fine if you don't. My problem is the core of this framework which is subject to be shared among the solutions. Node.JS needs a core, a compact framework, compact public API that is not dependent (at least this much) to the underlying JS engine / openssl version etc...

> FFI will be brought to io.js, which should make trivial addons doable straight from JavaScript. 

I see why it's needed and I see how it will be hard to maintain it on many platforms. I don't know if I should repeat compact vs rest of the API puzzle. 

> That being said, as there appears to be interest in spidernode or whatever, it seems reasonable to form a WG around that and let those who care for it work on it.

I would rather go for a Node.JS compact framework workgroup. I don't know if I'm the only person believes that we need something in the core of everything else. Once we have such a core, rest is a customization around it.

**Comment by imyller** [comment]
:+1: for this process, but I'm trying to (over)-simplify this a bit:

My vote goes to solution that offers both:

a) Compact API, JS engine independent, as a way of calling native compiled libraries. Forget the Local<> stuff and V8 variable types etc. Just plain native types. This excludes a lot functionality, but that's the point. Maintain this compact API in the core.

b) Full native API that we now use directly or with NAN for those who need/want to have full access to and from JS engine. 

I think many packages writing their native addons in C/C++ would just need simple "binding" between their native codebase and JS; not the overly fancy V8 facade, data types etc. 

If you look at most of the published native addon packages, they have clean native C/C++ code written without any consideration for V8/JS and then a single "node_binding.cc" that is the necessary evil to bind native world with the JS .. and its usually the only thing that breaks with core upgrades.

**Comment by pittek** [comment]
@imyller couldn't agree more.

**Comment by Venemo** [comment]
Just a thought about C++ and ABI compatibility. If the guys at Qt and KDE can do it, it's not impossible. Let's not rule out C++ yet. Here's what they have to say about it: https://techbase.kde.org/Policies/Binary_Compatibility_Issues_With_C++ — there are some do's and don'ts in there as well. If you use the “d-pointer” technique and treat virtual functions with care, you don't have much to worry about.

**Comment by geoffkizer** [comment]
Hey folks, I’m Geoff Kizer from the Chakra team at Microsoft.

Re @imyller:

> I think many packages writing their native addons in C/C++ would just need simple "binding" between their native codebase and JS; not the overly fancy V8 facade, data types etc.

This sounds like a simple FFI mechanism.  Is that what you meant?  

The nice thing about a simple FFI mechanism is that it makes the native code completely independent of the JS engine API.  As the V8 API changes, you only need to update the FFI implementation; users of the FFI don’t need to change.

This is basically what the CLR does.  Pretty much all native code is invoked through P/Invoke, including performance critical stuff like System.Net.

There’s some discussion of FFI for Node [here](https://github.com/joyent/node/issues/8285), but I’m not sure where it stands currently.

**Comment by trevnorris** [comment]
@geoffkizer Though, to be honest, I doubt the FFI bindings would be fast enough to meet your needs. I've tried some simple operations and it shows considerable overhead compared to directly making the call to C++.

**Comment by geoffkizer** [comment]
@trevnorris 
> Though, to be honest, I doubt the FFI bindings would be fast enough to meet your needs. I've tried some simple operations and it shows considerable overhead compared to directly making the call to C++.

I assume you're referring to node-ffi here?  I'd be curious to see what you tried and what the perf results were.

Perf is always a concern with an FFI.  It's easy to make the FFI slow, much harder to make it fast -- but not necessarily impossible.  The CLR's P/Invoke seems to do just fine here.  And I believe other languages (Java, Python) rely on an FFI for similar functionality.

I haven't looked at node-ffi in any detail, so I don't know how well it performs for particular scenarios.  Maybe it can be optimized further.  Or maybe a different (but still FFI-based) approach would be better for performance.

**Comment by geoffkizer** [comment]
Yeah, node-ffi has a ton of overhead.

Simple "add two ints" function, 1 million iterations:
Native C++ call: 17 ms
node-ffi to external library: 62535 ms

So using node-ffi seems like a non-starter.

I still think a "simple C/C++ binding" model would be useful.  Maybe we could do something like this with C++ templates, as @bnoordhuis suggested on the FFI thread...

**Comment by TooTallNate** [comment]
FWIW, the newly proposed `ffi` module in https://github.com/nodejs/io.js/pull/1865 is a lot more low-level of an API, and you could probably squeeze some performance out (or at least less memory usage) by using the new API directly.

**Comment by geoffkizer** [comment]
Here's an example of how we could use C++ templates to implement a simple native binding mechanism.

Consider a trivial native function, one that adds two integers and returns the result.  You’d implement it like this:

    void AddMethod(const FunctionCallbackInfo<Value>& args) {
      int32_t x = args[0]->Int32Value();
      int32_t y = args[1]->Int32Value();
      int32_t result = x + y;
      args.GetReturnValue().Set(result);
    }
    
    void Initialize(Handle<Object> exports) {
      NODE_SET_METHOD(exports, "add", AddMethod);
    }

Most of this is just glue code, to deal with extracting and converting arguments and setting the return value.

With templates, we can make all this glue code automatic, based on the native function signature.  Here’s what this would look like:

    int32_t AddMethod(int32_t x, int32_t y) {
        return x + y;
    }
    
    void Initialize(Handle<Object> exports) {
      NODE_SET_METHOD(exports, "add", BindNativeFunction<AddMethod>);
    }

Here, all the V8 “glue” is gone, replaced with simple native types (int32_t here).  The magic is in the “BindNativeFunction” template used in the NODE_SET_METHOD macro.  It understands how to marshal int32_t to/from JS and provides the necessary glue automatically, based on the argument and return types.

Because this all happens at compile time, the performance is virtually identical.  For 1 billion iterations of each:
Original implementation: 14929 ms
“Template binding” approach: 15074 ms

That’s a difference of ~1% on a completely trivial function.  (Likely, one of the helper functions isn’t getting inlined fully; I haven’t checked the actual code.)

There are several nice things about this approach.  To begin with, it means you don’t have to write all that glue code – just use the right types in your signature and marshalling happens automatically.

More importantly, though, it isolates the module code from any V8 API changes.  If “Int32Value()” suddenly becomes “GetInt32Value()”, the only thing that needs to change is the template code itself.

Obviously this is a very simple example, and we'd need to flesh this out a lot more to support real world scenarios.  But I think there's a lot of potential goodness here.

**Comment by domenic** [comment]
This is pretty similar to how Chrome's JS bindings occur actually, although they have a WebIDL intermediation layer (which has its own pluses and minuses...). It does indeed mostly protect us from V8 API changes, in that most of the code doesn't need to use V8 types at all, as long as we update the binding generators.

**Comment by trevnorris** [comment]
I like how sleek that looks, but seems to depend on the call being monomorphic. Have a way around that?

**Comment by geoffkizer** [comment]
@trevnorris: do you have a particular example in mind?

**Comment by trevnorris** [comment]
Any time a call is made directly from user facing API to native it's a possibility. To get around that we'd need to pad all calls by going through JS first. Fundamentally that's probably the better approach, but not how it's always done now.

Example would be Buffer#copy(). Direct call to native and has optional parameters.

**Comment by geoffkizer** [comment]
Yes, I'd suggest doing the processing in JS first and call out to a monomorphic native function.  As you say, this is generally a better approach.  

Also, we could support native argument types like optional<T> if that's generally useful.  That would help with Buffer#copy.

**Comment by trevnorris** [comment]
if we take this approach, might as well just always go through JS first.

using this approach will it throw or abort if there's an argument type mismatch?

**Comment by geoffkizer** [comment]
Throw on argument type mismatch, I think.

**Comment by domenic** [comment]
In case it is helpful here is a very simple example of the binding generation in Blink

- [What we write](https://code.google.com/p/chromium/codesearch#chromium/src/third_party/WebKit/Source/core/dom/DOMRectReadOnly.h&cl=GROK&l=19&gsn=DOMRect&rcl=1433952319): `double x() const { return m_x; }`
- [What gets generated](https://code.google.com/p/chromium/codesearch#chromium/src/out/Debug/gen/blink/bindings/core/v8/V8DOMRectReadOnly.cpp&sq=package:chromium&l=32): the usual `static void xAttributeGetter(const v8::FunctionCallbackInfo<v8::Value>& info)` with conversion and extraction stuff inside it.
- [WebIDL glue](https://code.google.com/p/chromium/codesearch#chromium/src/third_party/WebKit/Source/core/dom/DOMRectReadOnly.idl&sq=package:chromium&l=14): not really necessary for this simple case, although the Web IDL spec does give the semantics for converting from arbitrary JS input to something that fits into a C++ double.

**Comment by obastemur** [comment]
Depending on statically defined functions or templates may lower the flexibility of JavaScript kind of method declarations. 
 - No definition of the number of arguments
 - No return type defined

Let me share you a sample code from JXcore's native interface. This is JavaScript engine independent.

```
void sum(JXValue *args, int argc) {
  if ( argc < 2 || !JX_IsDouble(0) || !JX_IsDouble(1) ) {
     return; // or throw an exception..
  }
  double a = JX_GetDouble(0);
  double b = JX_GetDouble(1);

  JX_SetDouble(args+argc, a+b);
}
```
This interface also covers;
 - Playing with JS Objects  / Array(Get, Set sub properties)
 - Re using / referencing JS side functions

I borrowed the root of this interface from SpiderMonkey while keeping V8 in mind. 

JXValue structure given below is intended to be an engine independent data store;
```
enum _JXType {
  RT_Int32 = 1,
  RT_Double = 2,
  RT_Boolean = 3,
  RT_String = 4,
  RT_Object = 5,
  RT_Buffer = 6,
  RT_Undefined = 7,
  RT_Null = 8,
  RT_Error = 9,
  RT_Function = 10
};

struct _JXValue {
  void *com_;
  bool persistent_;
  bool was_stored_;

  void *data_;
  size_t size_;
  JXValueType type_;
};
```

I don't see this API is changing / breaking because of an update with V8 or SpiderMonkey. However, this structure is new and needs some updates (i.e. byte ordering) or can be smarter. Yet it does the most of the task needed so far. 

I hope the interface for node.js will be abstracting v8 types.

**Comment by trevnorris** [comment]
Problem I see with creating argument type specific methods is that the number of types is very large: https://github.com/v8/v8-git-mirror/blob/4.4.65/include/v8.h#L1647-L1894

**Comment by geoffkizer** [comment]
@domenic: Interesting, thanks.  IE uses something similar internally.

I think a template-based approach and a WebIDL-based approach are pretty similar; both provide automatic marshalling to/from native types with minimal performance overhead, and both hide the engine API details from the native code author.

The nice thing about the template-based approach is that the mapping is defined and generated by the native function signature itself.  No separate WebIDL file, no codegen step, etc.

WebIDL may enable more expressivity in the mapping.  I'm not that familiar with WebIDL and so I don't know if this matters much.  It might be interesting to see how much of WebIDL semantics we could implement using templates.

**Comment by domenic** [comment]
I think it does give more expressivity, but at the cost of a lot of obtuse behavior. I wouldn't really want to use Web IDL itself. The template thing sounds cool and hopefully will work if we wrap up anything complicated (like options objects) into a JS wrapper.

**Comment by geoffkizer** [comment]
@obastemur: I agree that a sufficiently rich engine abstraction API would be more flexible.

I think the interesting question is, how often do you actually need/want this flexibility?

My intuition is that you can get quite far with a simple native binding plus some JS code in front to make the object model pretty.  And unless you need the full power and flexibility of the engine abstraction API, you’re better off just using the simple native bindings.

For example, I think many of the internal native modules (e.g. tcp_wrap) would work well with a simple native binding; these are already fairly low-level native wrappers and the kinds of conversions they do are straightforward.  They don’t need the full flexibility of an engine abstraction API; they just need a simple mechanism to bind native functions into JS.

That said, it’s hard to know for sure without writing much more code.

**Comment by obastemur** [comment]
@geoffkizer Agree, the flexibility part is not that critical. Speaking of critical part, I could pick a basic list randomly from the same TCPWrap example.

 - Ability to interact with a JS Object (Create / Use / Return)
 - Save a JS function reference so you can call it later async. 
 - Wrap / Unwrap mem. reference
 - Return type ? Thrown Exception ? Null or Undefined?

Nice to have;
 - 'this' instance reference object
 - Manage persistence / GCRooting manually

Implementing such an interface is not that tricky and It's proven to work. I would like to know if there is a bottleneck that I'm not aware of ?

**Comment by trevnorris** [comment]
Just to clarity, are we still talking about an abstracted user facing API or an interface to abstract away V8 entirely?

**Comment by obastemur** [comment]
> Just to clarity, are we still talking about an abstracted user facing API or an interface to abstract away V8 entirely?

+1B for ` abstract away V8 entirely`

**Comment by trevnorris** [comment]
The list of needs to abstract away V8 entirely is much larger. A few of those are:

- Handle `Persistent<T>` and their weak callbacks.
- Be able to at minimum work with `ArrayBuffer`, Typed Arrays, `Array`, `Symbol`, `Function`, different numeric types and do string conversion from one byte, two byte or UTF-8.
- Throw exceptions from native.
- Receive notifications about specific events within the lifetime of a `Promise`.
- Starting in v3.x the ability to
  - Change an object's `__proto__` from native API.
  - Directly manipulate the memory contained within an `ArrayBuffer` or create a new one from existing memory.

These are things currently used in core, and I've attempted to keep it to essentials. Nothing too V8 specific. Like being able to have a persistent handle visitor that can walk the list of all `Persistent<T>` that have been assigned a wrapper class id.

Though loss of all the existing tooling for analysis, debugging and performance would be a big hit.

I'm not trying to poop on the parade. Just want to give a realistic view of what abstracting away the entire V8 API would look like for core.

**Comment by geoffkizer** [comment]
I think the goal is an abstracted user facing API.

I think the requirements for a user facing API vs the requirements for Node core are quite different.  As you list above indicates, trying to provide a full abstraction for Node core makes the problem much larger.

To me, the key observation here is that most modules don’t actually need anything like the full V8 API.  What they need is a much smaller set of functionality that allows them to marshal simple types between JS and native.  Do JS stuff in JS, do native stuff in native, and provide a simple, minimal, and efficient mapping between the two.

The “template binding” approach I’ve suggested is intended to do that.  That said, as @obastemur indicated above, this could also be done with a minimalist abstraction API.  Either way, the key is to keep this minimal so we don’t have to abstract the entire V8 API.

**Comment by obastemur** [comment]
> Either way, the key is to keep this minimal so we don’t have to abstract the entire V8 API.

I Agree. A lightweight, engine independent API is very critical for the future of node and it's ecosystem. If an addon requires a specific V8 etc. feature then it can use the current approach. 

Long story short, minimalist abstraction approach doesn't limit the capabilities. If you are using an engine specific feature, you better keep the addon updated to the latest changes (as usual). However, minimalistic API saves the majority of the modules from breaking. 

> Nothing too V8 specific.
@trevnorris  Thanks for the list. It may not sound a major difference but Persistent<..> is v8 (precisely v8 GC) specific. It does one basic thing in common. Keeps an 'Object' from garbage collection. The rest of it not that similar. 

For the rest of the non-v8 specific items, I didn't see any bottleneck or a too complicated API.

**Comment by trevnorris** [comment]
@obastemur The key need from `Persistent<T>` is not keeping objects from being GC'd. It's allowing us to know when the object is about to be GC'd. We have internal resources attached to objects, and those resources need to be cleaned up when the object is about disappear.

**Comment by obastemur** [comment]
@trevnorris Agree that's the key need but there is no matching feature on SM side. Lots of unrelated details though but if let's say the minimalistic API depends anything but `keeping objects from being GC'd` that would be engine specific.

**Comment by trevnorris** [comment]
@obastemur To clarify, are you saying that despite the different approaches, each VM has at least some way to achieve the functionality of being notified by the object when it's about to be GC'd?

**Comment by geoffkizer** [comment]
I don't know how SM works, but Chakra allows you to create finalizable objects which will notify you when they are about to be GC'd.  This is different than V8's MakeWeak mechanism, since it's specified as part of object creation instead of later.  But it provides the functionality that Node needs.

**Comment by trevnorris** [comment]
@geoffkizer Thanks for the insight. I can't think of a place within core where an object is made weak after the fact.

Another piece of functionality that is very heavily used in core, and by more than a few modules, is the ability to attach void pointers to the object for later access. Mostly this is used to attach class instances to the object so they can be retrieved later.

@obastemur Want to clarify one more thing you said. Sounded like the concept of preventing an object from being cleaned up by GC through native API isn't possible through other VMs. Or were you just saying it is done differently?

**Comment by obastemur** [comment]
> To clarify, are you saying that despite the different approaches, each VM has at least some way to achieve the functionality of being notified by the object when it's about to be GC'd?

If lets say an object is created on JS land, it doesn't have a kind of "on garbage collected" event listener assigned and there is no way to mark this flag later. 

SM lets you set the flags for an object that you've **created on the native side**. However, we did combine multiple approaches to track the JS side objects for GC time. I was experimenting with JSC and it's similar to SM than V8 on this part.

> Sounded like the concept of preventing an object from being cleaned up by GC through native API isn't possible through other VMs. Or were you just saying it is done differently?

Keeping an object from GC is kind of a common result. In details it's different. 

Let me share you one of the tricky devils. Here is a generic node scenario; 
 - Persistent/Rooted object is being garbage collected.  (event received)
   : The engine supposed to be in sweep phase.
 - HandleWrap receives the event and destructs the wrapped native class.  
    : This is not a problem. It doesn't share the memory with JS engine. 
 - This native class ended up triggering (object creation, code execution etc.) on sweep phase.  
    : We just did corrupt the memory ?

The minimalistic API may provide the infrastructure for keeping an object from GC. Raising an event for GC is also **possible** but it's just so delicate to locate properly. Giving away V8 interface as is may break with both future V8 versions and other engines. 

> I don't know how SM works, but Chakra allows you to create finalizable objects which will notify you when they are about to be GC'd.

@geoffkizer Is it possible to mark a JS side object finalizable ? Or it's defined on creation (similar to SM)

**Comment by jianchun** [comment]
> Is it possible to mark a JS side object finalizable ? Or it's defined on creation (similar to SM)

@obastemur Chakra just added a new API in Win10 release to mark any object (including JS side object) finalizable:  [JsSetObjectBeforeCollectCallback](https://msdn.microsoft.com/en-us/library/mt125488(v=vs.94).aspx). The callback can do things like destructing native wrapper, but attempt to create object / execute code will be rejected by Chakra runtime because of the same reason as you mentioned.

**Comment by trevnorris** [comment]
@obastemur Currently we only need to track JS objects made on the native side. The technical reasons have to do with creating an object that allows tracking a void pointer.

While technically the weak callback in V8 is called post-gc, we aren't doing any object creation or the like during that time anyway. So it shouldn't be a problem. The most important thing is that we can gain access to the void pointers that were attached to the object so we can do all necessary cleanup.

> Giving away V8 interface as is may break with both future V8 versions and other engines.

I haven't addressed anything specific to V8's implementation details. Instead what basic functionality is absolutely necessary in order to get node running. Like being able to be notified when an object has been cleaned up by GC, attach void pointers to an object's internal fields, or prevent an object from being collected by GC when it's no longer referenced in JS.

**Comment by bmeck** [comment]
I think we all need to remember that this is about providing a generic abstract set of Runtime behaviors that Node uses. If an engine does not support the behavior in the same manner as v8 whatever Runtime it provides for Node compatibility will just have to be documented. I have no desire for Node to try and satisfy all the implementation differences between engines. Documenting and specifying what Runtime behaviors we need can be done however, even if not all engines currently support the behavior.

**Comment by obastemur** [comment]
> I have no desire for Node to try and satisfy all the implementation differences between engines. Documenting and specifying what Runtime behaviors we need is doable however, even if not all engines currently support the behavior.

I would suggest you to check the changes to V8's GC over time. There are no thousands of ways for GC and not a long standing best one. This is a delicate place that may change from one version to another. Even if you have no interest for other JS engines, taking other engines into reference would help the final interface less vulnerable. 

> While technically the weak callback in V8 is called post-gc, 

Sweep or post sweep doesn't make much difference since it pauses the main thread. The problem is that the object is still reachable on event but it's memory location is vulnerable (for post sweep case). 

> we aren't doing any object creation or the like during that time anyway. 

Actually we do. There are scenarios to mention but saw this basic example on https://github.com/joyent/node/blob/master/src/node_object_wrap.h#L119-L130 While we are using it inside the node, addon developers may do the same. 

I'm optimistic. It may sound complicated but actually it's not. We can develop a minimalistic addon interface that will 
 - be easier to use (no handle<>, local<> etc. stuff) 
 - let developers benefit from other engines (Chakra, SM, Duktape, JSC..)
 - not that vulnerable to v8's or other engines API changes (they are all changing though)

Actually some of these engines are the only options on some platforms. Chakra on Windows ARM and SM/JSC on iOS. So, let native node addons reaching to other platforms.. node apps are already there.

**Comment by geoffkizer** [comment]
@obastemur: how do you get rid of handle<> and local<>?

**Comment by bmeck** [comment]
@obastemur 

>  I would suggest you to check the changes to V8's GC over time. There are no thousands of ways for GC and not a long standing best one. This is a delicate place that may change from one version to another. Even if you have no interest for other JS engines, taking other engines into reference would help the final interface less vulnerable.

I don't understand the implication here with respect to needing abstract behaviors to be listed for compatibility.

**Comment by trevnorris** [comment]
I feel like the proposal is moving from "let's create an abstraction layer from how node currently works" to "let's rethink the fundamental architecture to accommodate all possible variances". I'm skeptical that would fly with the TSC.

There's functionality that Node needs. Like knowing specific information about the life of an object. I remember a few years ago Ben attempted to implement node on top of SM, but found it wasn't possible to keep feature parity. Node won't sacrifice features to achieve an abstraction. The majority of the community doesn't care about the abstraction. We need to make the case why it's in their best interest to do so. Regardless of preventing V8 breakage, if features are gimped then it won't be welcome.

**Comment by kzc** [comment]
> a few years ago Ben attempted to implement node on top of SM, but found it wasn't possible to keep feature parity.

@trevnorris, what v8 feature(s) specifically? [JXCore](https://github.com/jxcore/jxcore) is a working Node fork on top of an engine abstraction API supporting both SM and V8. So it's clearly possible.

Are you suggesting that the node foundation would only entertain supporting non-v8 engines only if they were wrapped by a v8-compatible C++ API as Microsoft has done with Chakra? But even so, the v8 API is a moving target - hence this discussion to have an engine agnostic API wrapper.

**Comment by geoffkizer** [comment]
> The majority of the community doesn't care about the abstraction. We need to make the case why it's in their best interest to do so.

I agree.

Here's the case as I see it for a "simple native binding", based on templates or similar:

(1) Simplifies your native code.  No need to learn a JS engine API, V8 or abstracted or otherwise.  Just use the native types supported by the binding layer and everything works.  Easy to use, easy to get right.
(2) Isolates you from V8 API breakage.
(3) Makes it easier to support other JS engines.

Most module authors don't care much about (3).  They care about (1) and (2).  Done properly, I think this could be compelling enough to make them want to use it.

**Comment by trevnorris** [comment]
@kzc It was years ago, and most likely that whatever was lacking at the time now exists. And in terms of V8, I'm saying that no existing functionality can be sacrificed because some VM doesn't support it. The abstraction needs to allow all public APIs to continue working as they do today.

@geoffkizer Those are good points, and it sounds like you are more directly addressing native module compatibility. The discussion does seem to have diverged into whether all of core should be abstracted or just the module interface. Which are you looking for?

**Comment by obastemur** [comment]
> how do you get rid of handle<> and local<>?

@geoffkizer I've mentioned on my previous comments a bit. JXcore's embedding interface is JS engine independent. [API](https://github.com/jxcore/jxcore/blob/master/doc/native/Embedding_API_Details.md) The approach can be improved and used for native add-ons. 

>  I have no desire for Node to try and satisfy all the implementation differences between engines. 
>  I don't understand the implication here with respect to needing abstract behaviors to be listed for compatibility.

@bmeck What I was saying that; we better know the implementation differences among the engines, so we know what is minimalistic and less likely to be broken. I agree with the documentation part.

> I feel like the proposal is moving from "let's create an abstraction layer from how node currently works" to "let's rethink the fundamental architecture to accommodate all possible variances".

I thought the discussion was a minimalistic API for native addons.

**Comment by kzc** [comment]
> The abstraction needs to allow all public APIs to continue working as they do today

@trevnorris do you consider the (ever changing) v8 API to be part of Node's public API? Where is the line drawn for the public API?

**Comment by trevnorris** [comment]
@kzc I'm only considering the APIs we have direct control over. Don't care about the V8 API.

@obastemur This thread has gotten long, and wires were crossed somewhere. Thought that some of this discussion had shifted to removing V8 from core. Native modules is a much easier target. Thanks for clarifying.

**Comment by obastemur** [comment]
Agree it's been a long thread. Luckily, many experienced developers participating on this discussion. How about prepare and share a minimalistic API (coverage) that is compatible to the engines discussed here.  

We might have a better picture on the actual facts / problems. Moreover, we may see if it's reasonable or not. 

I can spend some time / or collaborate on the initial list upcoming days.

**Comment by geoffkizer** [comment]
@trevnorris:

> The discussion does seem to have diverged into whether all of core should be abstracted or just the module interface. Which are you looking for?

I think the pressing need is to abstract the module interface, so that's what I think we should focus on.

That's what I meant above re "abstracted user facing API"; I should have said "abstracted user facing module API" to be clearer.

**Comment by trevnorris** [comment]
For anyone that's participated, I'll be writing up a summary of what these abstractions are attempting to achieve on https://github.com/nodejs/io.js/issues/1993

Basically we're just in the process of properly forming a WG to handle these things.

**Comment by trevnorris** [comment]
For everyone's information, the API WG repo has been created. I'm still writing the initial README, but feel free to join if you wish to get involved: https://github.com/nodejs/api

I plan for all communication possible to be async. So submit issues and whatnot and we'll get things going.

**Comment by trevnorris** [comment]
Everyone, @obastemur was awesome enough to write up the first PR for the API WG. This has to do w/ the native API, and those that have participated in this thread are more qualified to give feedback from myself: https://github.com/nodejs/api/pull/4

**Comment by Ben007** [comment]
A bit outdated but just saw this discussion.I am trying to put up a cross platform native addon. I can't tell how hard it is to catch up with all the changes and differences. Looking forward to see a common API! 

> I think the pressing need is to abstract the module interface, so that's what I think we should focus on.

@geoffkizer exactly.

> @kzc It was years ago, and most likely that whatever was lacking at the time now exists. And in terms of V8, I'm saying that no existing functionality can can't be sacrificed because some VM doesn't support it. The abstraction needs to allow all public APIs to continue working as they do today.

@trevnorris nothing new with spider monkey.  @obastemur implemented the missing features into spider.
